sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/AccumulatedMetricResults.java

```diff
@@ -0,0 +1,325 @@
+      return
+          FluentIterable
+              .from(gauges.values())
+              .filter(matchesFilter(filter))
+              .transform(gaugeUpdateToResult())
+              .toList();
+    }
+
+    private Predicate<AttemptedAndCommitted<?>> matchesFilter(final MetricsFilter filter) {
+      return new Predicate<AttemptedAndCommitted<?>>() {
+        @Override
+        public boolean apply(AttemptedAndCommitted<?> attemptedAndCommitted) {
+          return MetricFiltering.matches(filter, attemptedAndCommitted.getKey());
+        }
+      };
+    }
+  }
+
+  protected Function<AttemptedAndCommitted<Long>, MetricResult<Long>>
+  counterUpdateToResult() {
+    return new
+        Function<AttemptedAndCommitted<Long>, MetricResult<Long>>() {
+          @Override
+          public MetricResult<Long>
+          apply(AttemptedAndCommitted<Long> metricResult) {
+            MetricKey key = metricResult.getKey();
+            return new AccumulatedMetricResult<>(
+                key.metricName(),
+                key.stepName(),
+                metricResult.getAttempted().getUpdate(),
+                metricResult.getCommitted().getUpdate());
+          }
+        };
+  }
+
+  protected Function<AttemptedAndCommitted<DistributionData>, MetricResult<DistributionResult>>
+  distributionUpdateToResult() {
+    return new
+        Function<AttemptedAndCommitted<DistributionData>, MetricResult<DistributionResult>>() {
+          @Override
+          public MetricResult<DistributionResult>
+          apply(AttemptedAndCommitted<DistributionData> metricResult) {
+            MetricKey key = metricResult.getKey();
+            return new AccumulatedMetricResult<>(
+                key.metricName(),
+                key.stepName(),
+                metricResult.getAttempted().getUpdate().extractResult(),
+                metricResult.getCommitted().getUpdate().extractResult());
+          }
+        };
+  }
+
+  protected Function<AttemptedAndCommitted<GaugeData>, MetricResult<GaugeResult>>
+  gaugeUpdateToResult() {
+    return new
+        Function<AttemptedAndCommitted<GaugeData>, MetricResult<GaugeResult>>() {
+          @Override
+          public MetricResult<GaugeResult>
+          apply(AttemptedAndCommitted<GaugeData> metricResult) {
+            MetricKey key = metricResult.getKey();
+            return new AccumulatedMetricResult<>(
+                key.metricName(),
+                key.stepName(),
+                metricResult.getAttempted().getUpdate().extractResult(),
+                metricResult.getCommitted().getUpdate().extractResult());
+          }
+        };
+  }
+
+  private void mergeCounters(
+      Map<MetricKey, AttemptedAndCommitted<Long>> counters,
+      Iterable<MetricUpdate<Long>> updates,
+      boolean isAttempted) {
+    for (MetricUpdate<Long> update : updates) {
+      MetricKey key = update.getKey();
+      AttemptedAndCommitted<Long> updated;
+      if (isAttempted) {
```

Comment:
Rather than a boolean, I think this could be a `Function<MetricUpdate<T>, AttemptedAndCommitted<T>>`. Then there are two versions of that function `AttemptedAndCommitted.fromAttemptedUpdate(defaultValue)` and `AttemptedAndCommitted.fromCommittedUpdate(defaultValue)`. That makes the code inside these merge methods cleaner (just calls the given function.